File: core/src/main/java/dev/bee/kanjianki/core/study/InkPoint.java
package dev.bee.kanjianki.core.study;

import java.util.Objects;

public final class InkPoint {
    public final float x;
    public final float y;
    public final long timestampMillis;

    public InkPoint(float x, float y, long timestampMillis) {
        this.x = x;
        this.y = y;
        this.timestampMillis = timestampMillis;
    }

    public InkPoint scaled(float width, float height) {
        return new InkPoint(x * width, y * height, timestampMillis);
    }

    @Override
    public boolean equals(Object other) {
        if (!(other instanceof InkPoint)) {
            return false;
        }
        InkPoint that = (InkPoint) other;
        return Float.compare(x, that.x) == 0
                && Float.compare(y, that.y) == 0
                && timestampMillis == that.timestampMillis;
    }

    @Override
    public int hashCode() {
        return Objects.hash(x, y, timestampMillis);
    }
}

File: writing-core/src/main/kotlin/dev/bee/kanjianki/core/study/InkPoint.kt
package dev.bee.kanjianki.core.study

import java.util.Objects

class InkPoint(
    @JvmField val x: Float,
    @JvmField val y: Float,
    @JvmField val timestampMillis: Long,
) {
    fun scaled(width: Float, height: Float): InkPoint = InkPoint(x * width, y * height, timestampMillis)

    override fun equals(other: Any?): Boolean {
        if (other !is InkPoint) {
            return false
        }
        return java.lang.Float.compare(x, other.x) == 0 &&
            java.lang.Float.compare(y, other.y) == 0 &&
            timestampMillis == other.timestampMillis
    }

    override fun hashCode(): Int = Objects.hash(x, y, timestampMillis)
}
